TensorLib/Basic.lean

```diff
@@ -2,4 +2,98 @@ namespace TensorLib
 
 def hello := "world"
 
+class TensorElement (a: Type) where
+  add: a -> a -> a
+  -- etc. enough to implement numpy ops
+
+inductive TensorData (a: Type) [TensorElement a] where
+| td_root (x: List a)
+| td_node (tl: List (TensorData a))
+
+-- Awkward... I'm sure this exists somewhere in the stdlib
+def sequence {a: Type} (x: List (Option a)): Option (List a) :=
+  match x with
+  | .nil => some .nil
+  | .cons none _ => none
+  | .cons (some y) ys =>
+    match sequence ys with
+    | none => none
+    | some ys' => some (.cons y ys')
+
+-- somehow prove termination here
+def Shape {a: Type} [TensorElement a] (x: TensorData a): Option (List Nat) :=
```

Comment:
Question: partial function or relation? Maybe we should implement both to get a feel for the two options.

Reply:
Ah didn't do this yet. Will add.

Reply:
Added.